AppController/djinn.rb

```diff
@@ -1372,21 +1457,70 @@ def get_property(property_regex, secret)
   #   A String containing:
   #     - 'OK' if the value was successfully set.
   #     - KEY_NOT_FOUND if there is no instance variable with the given name.
+  #     - NOT_READY if this node is not shadow, and cannot talk to shadow.
   #     - BAD_SECRET_MSG if the caller could not be authenticated.
   def set_property(property_name, property_value, secret)
     return BAD_SECRET_MSG unless valid_secret?(secret)
+    if property_name.class != String or property_value.class != String
+      Djinn.log_warn("set_property: received non String parameters.")
+      return KEY_NOT_FOUND
+    end
 
-    Djinn.log_info("Attempting to set @#{property_name} to #{property_value}")
+    unless my_node.is_shadow?
+      # We need to send the call to the shadow.
+      Djinn.log_debug("Sending set_property for #{appid} to #{get_shadow}.")
+      acc = AppControllerClient.new(get_shadow.private_ip, @@secret)
+      begin
+        return acc.set_property(property_name, property_value)
+      rescue FailedNodeException
+        Djinn.log_warn("Failed to forward set_property call to #{get_shadow}.")
+        return NOT_READY
+      end
+    end
 
-    name_with_at_sign = "@#{property_name}"
-    begin
-      instance_variable_set(name_with_at_sign, property_value)
-    rescue NameError
-      Djinn.log_info("Failed to set @#{property_name}")
+    Djinn.log_info("Received request to change #{property_name} to #{property_value}.")
+    opts = {}
+    opts[property_name] = property_value
+    newopts = check_options(opts)
+
+    # If we don't have any option to set, property was invalid.
+    if newopts.length == 0
+      Djinn.log_info("Failed to set property '#{property_name}'.")
       return KEY_NOT_FOUND
     end
 
-    Djinn.log_info("Successfully set @#{property_name} to #{property_value}")
+
+    # Let's keep an old copy of the options: we'll need them to check if
+    # something changed and we need to act.
+    old_options = @options.clone
+    newopts.each { |key, val|
+      # We give some extra information to the user about some properties.
+      if key == "keyname"
+        Djinn.log_warn("Changing keyname can break your deployment!")
+      end
+      if key == "max_memory"
+        Djinn.log_warn("max_memory will be enforced on new appservers only.")
+      end
+      if key == "min_images"
+        unless is_cloud?
+          Djinn.log_warn("min_images is not used in non-cloud infrastructures.")
```

Comment:
Since cloud/cluster distinction is going away, I suggest we change this to something like:

`"[...] not used in agent-managed infrastructures"`
or
`"[...] not used in managed infrastructures"`